### ISSUE TITLE
Startup/restart performance degradation (10s legacy image vs 30-50s new image)

### ISSUE THREAD

**Issue by outcoldman** [splunk/docker-splunk#10]
The old image takes ~10 seconds to start Splunk. The new image takes 30 seconds. Unpacking splunk.tar.gz on every start takes about 20 seconds additionally.

Because containers are running on a writable container layer without persistence, you will find these additional ~20 seconds not only on `docker run,` but also `docker restart`. 

I tested on MacBook Pro with NVMe SSD, I assume that on HDD/SSD that will even more visible. 

You mentioned that you reduced 30% of the image by keeping Splunk gzipped. But that does not provide any benefits because: 

* docker takes more time to start the new image,
* only in rest you can see that image takes less space than the old image, which does not give you any benefits, considering that if you downloaded the image to the host, you probably want to start it
* in transition docker always sends gzipped layers, so size/time of pull/push will be approximately the same

**Comment by bb03** [comment]
+ @nwang92 @lephino

**Comment by lephino** [comment]
While size is a small portion of the benefits to the container, the much larger ones have to do with upgrades and volume mounting styles in the new container.  The new style will support a user mounting ANY volume under splunk in any directory that they wish.  A user will also be able to use 1 stateful volume on say a version 7.2 and know that in 7.2.1 that volume will have any upgrade processes required ran on those custom volume mounts.  While the costs of spinning up a new container are heavier on CPU than the old, the trade for supportability and upgradability greatly out way the start time.  We are always discussing ways to keep all the functionality and lower start times, if you happen to think of one, don't hesitate to get us the information.  

Was there an issue we needed to address in this issue request?  Or just a place to start an open dialog?  I just don't want to close the issue if I missed it.

**Comment by outcoldman** [comment]
> Was there an issue we needed to address in this issue request?

To be honest this is a blocking issue for me from switching to a new container. I cannot talk for everybody else. Just an additional 20 seconds for startup will increase the length of the tests and time that I need to bootstrap a new development environment. When people compare containers to VM - one of the main benefits is a startup time. For 20 seconds I can start a new VM and also run apt-get update && distr-upgrade on this VM. 

> The new style will support a user mounting ANY volume under splunk in any directory that they wish. 

Looking at https://github.com/splunk/docker-splunk/blob/master/splunk/debian-9/Dockerfile#L46, you define two volumes, which is basically a recommended way. I understand that with the new approach you can support more scenarios, and maybe mount whole `/opt/splunk`, but why? Even if you have this scenario, how often do you need it? Can we just find a way to work around this specific scenario?

**Comment by lephino** [comment]
As you're aware, when you're first creating volumes, they are empty and require some "copying" around to make them work.  As we move into the world of orchestrators updating the apps, you have to remember, this isn't about the "first" run of the app, it's the 2nd.  If you put 7.2.0 on, and then you want to move the data to 7.2.1, you're now requiring every single person to run the docker build with their named mounts attached (as now since the volumes exist, files won't be copied back to the volume).  Splunk is still a stateful app and requires an upgrade process to be ran on conf files and indexes.  This means, as we move to support, we need to make sure those FTR scripts are ran on any persistent data that exists.  Untarring the app at startup ensures that no matter what, the latest files that you started are always applied to the named mounts. and as you move to an orchestrator upgrade, that a container could be handed the same existent volumes, and migrate successfully.

**Comment by outcoldman** [comment]
@lephino correct, the upgrade issue was solved a long time ago https://github.com/splunk/docker-splunk-legacy/blob/master/enterprise/entrypoint.sh#L23, you can compare version inside of the `/opt/splunk/etc` and backup folder and copy it over only in case if that has been changed.

**Comment by lephino** [comment]
I didn’t say it wasn’t solved. I said it requires different interactions and different complexities and causes different issues. This issue can be solved in more ways as well.  I’m saying; at this time, 20s does not justify the added complexity or the support cost to handle that complexity. We will continue to investigate structures to decrease install time.  Our overall focus should be on that; total time till uptime.  Give me a realistic number that you’re saying a container must be up by instead of just disagreeing with implimentations.  You’re saying that we must have a container up and running in a standalone configuration in under 10s and 30s takes too long and a non-starter? Correct?

**Comment by outcoldman** [comment]
@lephino let me rephrase the issue to `Startup/restart performance degradation (10s legacy image vs 30-50s new image)`. I am a developer myself and had some experience with the Splunk running inside images. So instead of writing a ticket about the problem I experienced, I pointed to the reason why I believe I saw this issue.

At this point, I honestly do not care about the implementation details. At first, I thought that is a brilliant idea. But after running it, I realized that this caused performance degradation. 

Just for curiosity, I tried to run this image on `t2.micro` instance. On this EC2 instance, the difference is more significant: 50 seconds comparing to the 10 seconds(!). As you can see the instance type does not affect the old image but does affect the new image.

Let me put the list of issues I see:

1. Startup time on the first start. Can degrade the performance of testing environments where we run images very often. Maybe affects production environments on ASG with the Indexers and new SmartStore. 20 seconds could make a big difference. 

2. Restart time of the container. That affects the performance of the production environments. 20 seconds for nothing does not seem like a good idea, especially if you are trying to fix the production environment, but restarting the container takes 20 seconds just to untar the Splunk. 

3. UF. I assume the degradation here is smaller, considering that `splunk.tar.gz` is smaller. That I will speak for the customers, as I do not care about this use case. But in case of ASG - if your agent that forwards metrics and logs takes 20 seconds to start, not sure if customers will be happy.

**Comment by lephino** [comment]
1.) Yes spinning up and down often will impact performance and do understand what you're saying.  Also CPU speed directly influences how fast it can uncompress (just like on bare-metal) 
2.) How are you restarting the container?  Docker restart or equivalent?  There is a function in the entrypoint to restart and not eat this cost time without the container unpackage times.  We also had a start routine that we could back in that did NOT untar splunk.  We removed it based on usage patterns and customer feedback.  In some instances customers said they modified "default" files in the etc / system / var directories, and realized afterwards that they didn't back them up or anything.  They wanted the container to auto-repair splunk files that were modified, and set them back to the expected state.  It makes this a challenging issue to solve.
3.) I think we can collect timing metrics and see if people really do have an issue with a larger start time.  I'll leave that up to @mikedickey and @amitmookerjee and we can adjust it as they see fit.  If this is a problem, we'll address it and get it reduced.

**Comment by outcoldman** [comment]
About (2) - I am talking about `docker restart` (or `docker stop` and `docker start`) not a Splunk restart, which I do quite often as well.

**Comment by lephino** [comment]
yeah for 2) you can do a work around with docker exec calling entrypoint.sh with a restart. 

```bbingham$ docker exec ba79278bc315 entrypoint.sh restart```

and you'll see in the main container:

```2018-10-15 21:58:18.179 +0000 Interrupt signal received
2018-10-15 21:58:34.243 +0000 splunkd started (build 8c86330ac18)
```

this one we're already talking about internally and feeling out how many people use stop / start and what the right approach is  (and what should be expected out of splunk).  Adding this to an example of why "untarring" to fix mistakes people use has negative effects.  @mikedickey @amitmookerjee

**Comment by outcoldman** [comment]
@lephino btw, just another proof why there is no 30% improvement of the image size, take a look at

https://hub.docker.com/r/splunk/splunk/tags/ - 7.2 - 394 Mb
https://hub.docker.com/r/outcoldman/splunk/tags/  - 7.2.0 - 407 Mb

It because images are transferred in a compressed format. 

> `entrypoint.sh restart`

That feels like a solution of "you are holding it wrong". I need to restart the container because I could stop Splunk or the host where I have Docker.

**Comment by lephino** [comment]
@outcoldman that's the compressed tag on dockerhub?  Do a docker images on a host you download the image to.

and the entrypoint I mentioned is something we're addressing.  I said that's a workaround.

**Comment by outcoldman** [comment]
@lephino I would assume on the host images are unpackaged, for the same reason why it will be good to unpack the splunk.tar.gz as well. To reduce the startup time. Image if every image in docker will require seconds to start? That will kill the whole point of containers.

**Comment by arctan5x** [comment]
@outcoldman Take a look at Mike's open PR. I think he is staging different container layers to address this problem. Since this issue has been open for a while, I am going to close this. If the issue arises to be critical, please reopen and escalate the issue.
